### ISSUE TITLE
Word: Unstable cross-references causing Word compare to highlight all links as changes

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/metanorma-iso#881]
I am comparing two Metanorma generated Word documents generated a few days apart.

These sections actually have no change, but are highlighted as changed. It seems that the Metanorma generated Word may not use stable IDs causing Word to think the links are different.

<img width="773" alt="Screen Shot 2022-12-08 at 11 47 37 PM" src="https://user-images.githubusercontent.com/11865/206492757-d487bf0e-441d-4aac-821e-4485336daf46.png">

**Comment by opoudjis** [comment]
I cannot replicate this with a simple eref crossreference:

````
= X
:no-pdf:
:stem:
:title-intro-en: Hello
:stage: 40
:docsubtype: vocabulary
:subdoctype: vocabulary
:requirements-model: ogc
:publisher: ISO;IEC

== Clause

<<guid>>

[bibliography]
== Normative references

* [[[guid,Globally Unique IDentifier (GUID)]]], span:title[2.5.5 Globally Unique Identifiers (GUIDs)], span:date.accessed[2013-01-17], Available from World Wide Web, span:uri[http://msdn.microsoft.com/en-us/library/cc246025.aspx] span:type[booklet]
````

Indeed, given the underlying markup,

````
<a href="#guid1">Globally Unique IDentifier (GUID)</a>
````

—the only way I was able to replicate this was by changing the reference anchors for the references. 

````
* [[[guid,....]]]

to 

* [[[guid1,...]]]
````

If you've changed the reference anchors, well, I don't see what I can do, short of globally replacing all user-supplied anchors with calculated anchors in all output. Which strikes me as utterly overkill. In fact, user supplied anchors are going to be more stable than the alternative, since they are independent of the content of the references.

I've just compared ISO-19115-3 with a version I generated two days ago. The links were the same: all that has changed was the markup fixed I'd introduced to tables globally.

Rather than me jumping at shadows, I request that you send me the documents you're comparing. I'm not replicating your results with documents generated under reasonable conditions (not completely the anchors of references), so I cannot tell you why you're getting those results. (And if you *are* changing the anchors of references, the result is no surprise: you're the one introducing the instability in the links, not Metanorma.)

**Comment by opoudjis** [comment]
I've compared the two documents you've sent me.

There is no content change around the links, and if I choose to display only content and not formatting in track changes, the only differences indicated between the documents are a bunch of stripped carriage returns, smartquotes in code snippets that are now straightened, a few image changes, and the removed 0 label for Introduction.

I am not getting delete and re-add links:

Content-only changes:

<img width="1079" alt="Στιγμιότυπο 2022-12-22, 20 19 11" src="https://user-images.githubusercontent.com/1911039/209101297-ebf72c96-f643-4908-928f-bda23cfafdba.png">

vs.

<img width="1079" alt="Στιγμιότυπο 2022-12-22, 20 20 03" src="https://user-images.githubusercontent.com/1911039/209101308-611b91ea-5193-4c6e-ae98-a19b608f0bde.png">

If I turn on formatting for track changes, I do get spurious change notifications, but only for a couple of the links, not for all of them. I will not that I get those spurious formatting changes in native Word documents I edit as well:

<img width="833" alt="Screen Shot 2022-12-22 at 20 24 14" src="https://user-images.githubusercontent.com/1911039/209102250-c771ffb8-4138-485f-9e25-5e0a9e3bb9be.png">

If you're seeing more changes than that, well, I'm not replicating that.

I've saved the two documents as DOCX and repeated. Same behaviour.

**Comment by opoudjis** [comment]
There is a random number inserted in DOC in bookmarks, and that is why Tables of Content will always be treated as different, no matter that their content might be identical. I could try to prevent that, but 

(a) I don't think it would work, 
(b) I don't think it would persist—the only way to get accurate page numbers in a ToC is to rebuild it, and rebuilding the ToC is going to insert new bookmark numbers anyway
(c) I don't think there's any points—people expect ToC to be different each time anyway
(d) it does not address this issue.

**Comment by opoudjis** [comment]
<img width="833" alt="Screen Shot 2022-12-22 at 20 35 14" src="https://user-images.githubusercontent.com/1911039/209104384-c01e5b0e-eca3-4527-b773-5a73fc222e1f.png">

And I'm pretty sure I'm not doing anything dodgy here.

I can't replicate this, and I don't see any difference in hyperlinking between the two documents such that I can action. Request to close this.

**Comment by ronaldtse** [comment]
@opoudjis I can verify what you are seeing with the 2 copies I sent. In any case I have sent you the version that was edited by the TC 211 secretariat, SIS. The changes may be real but I did not want to take chances here.

**Comment by opoudjis** [comment]
The edited version is of course an edited DOCX version, which means that Microsoft might be introducing all sorts of craziness there. 

Spurious  changes are happening to every single hyperlink.

If I compare the MHT export of the received DOCX with the sent MHT, I get:

* old: `<p class=ForewordText><a name="_eed71995-9c78-dd68-34e3-a8563521815d"></a><span
lang=EN-US style='mso-ansi-language:EN-US'>The procedures used to develop this
document and those intended for its further maintenance are described in the
ISO/IEC Directives, Part 1. In particular the different approval criteria
needed for the different types of ISO documents should be noted. This document
was drafted in accordance with the editorial rules of the ISO/IEC Directives,
Part 2 (see <a href="http://www.iso.org/directives.html"><span lang=EN>www.iso.org/directives</span></a>).<o:p></o:p></span></p>`
* new: `<p class=ForewordText><a name="_eed71995-9c78-dd68-34e3-a8563521815d"></a><span
lang=EN-US>The procedures used to develop this document and those intended for
its further maintenance are described in the ISO/IEC Directives, Part 1. In
particular the different approval criteria needed for the different types of
ISO documents should be noted. This document was drafted in accordance with the
editorial rules of the ISO/IEC Directives, Part 2 (see <a
href="http://www.iso.org/directives.html"><span lang=EN>www.iso.org/directives</span></a>).</span></p>`

The old document does not have language tags at all. The actual hyperlink text is exporting to MHT identically, including all its markup.

My surmise is that somewhere in the save chain (although I cannot replicate it), Word is adding `<span lang=EN-US style='mso-ansi-language:EN-US'>` some places, just `<span lang=EN-US>` in others, and is getting confused about the discrepancy. So adding `<span lang=EN-US style='mso-ansi-language:EN-US'>` everywhere there is text *might* prevent this.

But this is a massive change, it may well backfire, and it may not be what is at issue: when I compare the documents, Word tells me that the *Swedish* document is in Australian English, meaning that the *Swedish* document is ignoring all language tagging, and reverting to the system default language. That's inconsistent with the hypothesis that DOCX is more specific about language tagging than the original.

I've generated a variant document with the same language tagging as the Swedish document has, and compared the DOCX of the two documents, that differ only in language tagging. I did not replicate the behaviour. I set the paragraph language explicitly to Australian English. It recognised that change, but did not alter the hyperlink text.

The behaviour you're seeing with the Swedish document happens the *first* time save the variant file to DOCX and compare them. It does not happen when I recompare the files I've just compared.

The issue DOES go away when you save the source document as DOCX, and compare DOCX  to DOCX. Suggest you do that as a matter of course.
